backend/regalloc/regalloc_split_state.ml

```diff
@@ -106,6 +107,139 @@ end = struct
     !destructions_at_end, definitions_at_beginning
 end
 
+(* Optimizes `destructions_at_end` by avoiding the repeated spill of a constant.
+   If a register is set only once, then all spills of that register dominated by
+   a spill of the same register can be deleted. *)
+module RemoveDominatedSpillsForConstants : sig
+  val optimize :
+    Cfg_with_liveness.t ->
+    Cfg_dominators.t ->
+    destructions_at_end:destructions_at_end ->
+    destructions_at_end
+end = struct
+  (* CR-someday xclerc for xclerc: move to Cfg_loop_infos? *)
+  let is_in_loop : Cfg_loop_infos.t -> Label.t -> bool =
+   fun loops label ->
+    Cfg_loop_infos.EdgeMap.exists
+      (fun _ (loop : Cfg_loop_infos.loop) -> Label.Set.mem label loop)
+      loops.loops
+
+  type set =
+    | Exactly_once
+    | At_least_twice
```

Comment:
nitpicking: it would be more accurate to say `At_most_once` and `Maybe_more_than_once`.
Here `set` value abstracts how many times a temporary is assigned at runtime, and the analysis ignores condition, so a reg that is mapped to `Exactly_once`  may in fact be never assigned, and a reg that is mapped to `At_least_twice` may be assigned only once (or not at all), if the loop always runs for one iteration (or never, resp).

Reply:
(Renamed the constructors which
were indeed confusing as neither
"static" nor "dynamic" in their
semantics.)